wrappers/binaries.go

```diff
@@ -103,21 +104,77 @@ func detectCompletion(base string) (string, completionMode) {
 	}
 }
 
-// AddSnapBinaries writes the wrapper binaries for the applications from the snap which aren't services.
-func AddSnapBinaries(s *snap.Info) (err error) {
-	var created []string
-	defer func() {
-		if err == nil {
-			return
-		}
-		for _, fn := range created {
-			os.Remove(fn)
+// findExistingCompleters returns a list of existing completers that were not created by us.
+//
+// Note: Only base names are returned and not the full paths of the completers.
+func findExistingCompleters(s *snap.Info, dir string) (existingCompleters []string, err error) {
+	for _, glob := range s.BinaryGlobs() {
+		completers, err := filepath.Glob(filepath.Join(dir, glob))
+		if err != nil {
+			return nil, err
+		}
+		for _, completer := range completers {
+			if osutil.FileExists(completer) && !dirs.IsCompleteShSymlink(completer) {
+				existingCompleters = append(existingCompleters, filepath.Base(completer))
+			}
 		}
-	}()
+	}
+	return existingCompleters, nil
+}
 
-	if err := os.MkdirAll(dirs.SnapBinariesDir, 0755); err != nil {
+// ensureDirStateGlobsWithKeep same as osutil.EnsureDirStateGlobs but also keeps passed files unchanged.
+//
+// It overwrites content by adding a self-reference to the file entries that need to be kept unchanged.
+func ensureDirStateGlobsWithKeep(dir string, globs []string, content map[string]osutil.FileState, keep []string) (changed []string, removed []string, err error) {
+	if len(keep) == 0 {
+		return osutil.EnsureDirStateGlobs(dir, globs, content)
+	}
+	if content == nil {
+		content = make(map[string]osutil.FileState, len(keep))
+	}
+	for _, file := range keep {
+		content[file] = osutil.FileReference{Path: filepath.Join(dir, file)}
+	}
+	return osutil.EnsureDirStateGlobs(dir, globs, content)
+}
+
+// ensureSnapBinariesWithContent applies snap binary content but keeps existing completers unchanged.
+func ensureSnapBinariesWithContent(s *snap.Info, binariesContent, completersContent, legacyCompletersContent map[string]osutil.FileState) error {
+	globs := s.BinaryGlobs()
+	// Ensure binaries
+	_, _, err := osutil.EnsureDirStateGlobs(dirs.SnapBinariesDir, globs, binariesContent)
+	if err != nil {
+		return err
+	}
+	// Ensure completers but keeps existing completers unchanged
+	existingCompleters, err := findExistingCompleters(s, dirs.CompletersDir)
+	if err != nil {
+		return err
+	}
+	_, _, err = ensureDirStateGlobsWithKeep(dirs.CompletersDir, globs, completersContent, existingCompleters)
+	if err != nil {
+		return err
+	}
+	// Ensure legacy completers but keeps existing completers unchanged
+	existingLegacyCompleters, err := findExistingCompleters(s, dirs.LegacyCompletersDir)
+	if err != nil {
+		return err
+	}
+	_, _, err = ensureDirStateGlobsWithKeep(dirs.LegacyCompletersDir, globs, legacyCompletersContent, existingLegacyCompleters)
```

Comment:
maybe it's unavoidable, but it feels confusing to me that we run the Ensure for both legacy and not but one level up we detect one or the other? does this remark mark sense?

Reply:
When I looked at it with fresh eyes, it does look confusing and gives the impression that we ensure both types. In reality only one ensure gets passed content and the other is nil. An ensure call to both is needed to clear the one that is not being used.
Maybe a clarifying comment is needed or a refactor to make the code readable?

Reply:
Yes, after thinking a bit more I understood the part about needing to clear the other usage. A comment for sure would help, another thing that might help is trying to change the signature to take a completion variant and only *one* set of completion content, and do the which is which of things to pass to the ensures inside this function, instead of one level up:

```
ensureSnapBinariesWithContent(s *snap.Info, binariesContent, completersContent map[string]osutil.FileState, completionVariant completionMode)
```

or something like, that might be an issue with the key(s) of completersContent so that it might need to be passed in a different form?

Reply:
updated, thanks! the code is definitely more readable/intentional now.